File: backend/app/schemas/block_storage.py
from pydantic import ConfigDict, BaseModel
from datetime import datetime
from typing import List, Optional

class UserRef(BaseModel):
    id: int
    full_name: str
    model_config = ConfigDict(from_attributes=True)

# --- Detail Schemas ---
class BlockStorageDetailBase(BaseModel):
    block_id: int
    storage_location: Optional[str] = None
    remark: Optional[str] = None

class BlockShortInfo(BaseModel):
    id: int
    block_no: int
    block_code: str    # ดึงจาก @property ใน Model
    accession_no: str  # ดึงจาก @property ใน Model

    model_config = ConfigDict(from_attributes=True)

class BlockStorageDetailCreate(BlockStorageDetailBase):
    pass

class RunShortInfo(BaseModel):
    id: int
    run_no: str
    model_config = ConfigDict(from_attributes=True)

class BlockStorageDetail(BlockStorageDetailBase):
    id: int
    run_id: int
    stored_at: datetime
    block: Optional[BlockShortInfo] = None
    # disposal
    discard_status: bool = False
    discard_at: Optional[datetime] = None
    discard_by: Optional[UserRef] = None
    run: Optional[RunShortInfo] = None

    model_config = ConfigDict(from_attributes=True)

class BlockDisposeRequest(BaseModel):
    detail_ids: List[int]

# --- Run Schemas ---
class BlockStorageRunBase(BaseModel):
    run_no: Optional[str] = None
    user_id: int
    remark: Optional[str] = None

class BlockStorageRunCreate(BlockStorageRunBase):
    pass

class BlockStorageRun(BlockStorageRunBase):
    id: int
    started_at: datetime
    finished_at: Optional[datetime] = None
    details: List[BlockStorageDetail] = []
    operator: Optional[UserRef] = None
    discard_status: bool = False
    discard_at: Optional[datetime] = None
    discard_by: Optional[UserRef] = None

    model_config = ConfigDict(from_attributes=True)

class BlockStorageDisposeRequest(BaseModel):
    run_ids: List[int]

class BlockStorageDetailUpdate(BaseModel):
    storage_location: Optional[str] = None
    remark: Optional[str] = None

class BlockStorageRunCreateBatch(BaseModel):
    user_id: int
    items: List[BlockStorageDetailCreate] # รายการตลับที่จะจัดเก็บพร้อมกัน
    remark: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/blockStorage.ts
export interface BlockStorageBatchPayload {
  user_id: number;
  items: { block_id: number; storage_location?: string; remark?: string }[];
  remark?: string;
}

export interface BlockStorageDetailResponse {
  id: number;
  run_id: number;
  block_id?: number;
  storage_location?: string;
  remark?: string;
  stored_at?: string;
  block?: { id?: number; accession_no?: string; block_code?: string };
  run?: { id: number; run_no: string };
  discard_status?: boolean;
  discard_at?: string;
  discard_by?: { id: number; full_name: string };
}

export interface BlockStorageRunResponse {
  id: number;
  run_no: string;
  started_at: string;
  finished_at?: string;
  user_id: number;
  operator?: { id: number; full_name: string };
  remark?: string;
  details: BlockStorageDetailResponse[];
  discard_status?: boolean;
  discard_at?: string;
  discard_by?: { id: number; full_name: string };
}

export interface PendingStorageBlockNode {
  key: string | number;
  id: number;
  code: string;
  isCase: boolean;
  is_decal?: boolean;
  children?: PendingStorageBlockNode[];
}

export interface ScannedStorageBlock {
  id: number;
  code: string;
  accession_no: string;
  storage_location?: string;
  scannedAt: string;
}
